client/server/bundler/sections-loader.js

```diff
@@ -52,7 +60,11 @@ const loader = function() {
 	const options = getOptions( this ) || {};
 	const { forceRequire, onlyIsomorphic } = options;
 	let { include } = options;
-	let sections = require( this.resourcePath );
+
+	// Build all the sections during the build step.
+	let sections = forceRequire
+		? require( this.resourcePath )
+		: require( this.resourcePath ).filter( utils.filterSections );
```

Comment:
This section-filtering code reads from `config`, depends on `CALYPSO_ENV`, and runs at build time, as part of a webpack plugin/loader.

I think it would be better to postpone the section filtering to runtime. I.e., generate only one production bundle that can run as Calypso (in any environment like `stage` or `horizon`) or Jetpack Cloud, depending on how it's run:
```
CALYPSO_ENV=jetpack-cloud-stage node build/bundle.js
```

Is there an agreement on this? Or are there important reasons why there should be two bundles generated?

I can suggest ways to do the section filtering at runtime if we agree on the premise.

Reply:
There are a couple webpack configs. This uses all the sections when the `webpack.config.node.js` is run and it uses sections from the config when `webpack.config.js` is being used.  

Another place where we are doing the section filtering is on pages/index.js see https://github.com/Automattic/wp-calypso/pull/40022/files#diff-13202bbd2bfb14a1570c13babd9e8444R825 

Is this what you mean by doing filtering at runtime?

Reply:
The `webpack.config.node.js` config is used to build the server that runs with Node.js. This Express.js-based HTTP server does two things related to sections:
- for routes like `/stats/*`, makes sure that the JS chunks that are required to load Stats are included as `<script>` tags in the HTTP response. You can see this in action when opening `https://wordpress.com/stats/day` in browser and inspecting the HTML document you got in response:

<img width="607" alt="Screenshot 2020-03-11 at 11 31 36" src="https://user-images.githubusercontent.com/664258/76407658-fded1b00-638b-11ea-8189-10b15a66a353.png">

There's the `entry-main` chunk, there's `stats` and also all the shared chunks. The server knows what to send because each section has its own server-side handler.

- for isomorphic sections like `themes` or `log-in`, the server renders the complete HTML markup of the page.

`webpack.config.js` is used to build the JS bundle that runs in the browser. At this moment, because the section filtering is done at build-time, separate builds are required for Calypso and Jetpack Cloud.

Ideally, we could run just one Docker image for all and just set the `CALYPSO_ENV` environment variable to choose which app and environment exactly we want to run.

Reply:
In https://github.com/Automattic/wp-calypso/pull/40022/commits/5649277c2e05c528e8800e41c3e8df2706b2f31d I implemented the runtime filtering: `section-loader` will build all sections (it adds some generated code to the `sections.js` module and leaves it intact otherwise), and the code (both server and client) will the filter it at runtime according to the currently active `CALYPSO_ENV`.